subsys/net/ip/ipv4_autoconf.c

```diff
@@ -0,0 +1,280 @@
+		break;
+	case NET_IPV4_AUTOCONF_ANNOUNCE:
+	case NET_IPV4_AUTOCONF_CONFIGURED:
+		if (iface->ipv4.conflict_cnt == 1) {
+			/* defend IP */
+			send_announcement(iface);
+		} else {
+			/* unset host ip */
+			if (!net_if_ipv4_addr_rm(
+				    iface, &iface->ipv4.requested_ip)) {
+				NET_DBG("Failed to remove addr from iface");
+			}
+			/* restart probing after second conflict */
+			net_ipv4_autoconf_start(iface);
+		}
+
+		break;
+	default:
+		break;
+	}
+
+	return NET_DROP;
+}
+
+static void ipv4_autoconf_timeout(struct k_work *work)
+{
+	struct net_if *iface =
+		CONTAINER_OF(work, struct net_if, ipv4.ipv4_autoconf_timeout);
+
+	if (!iface) {
+		NET_DBG("Invalid iface");
+		return;
+	}
+
+	switch (iface->ipv4.state) {
+	case NET_IPV4_AUTOCONF_PROBE:
+		/* schedule next probe */
+		if (iface->ipv4.probe_cnt <=
+		    IPV4_AUTOCONF_PROBE_NUM - 1) {
+			send_probe(iface);
+			break;
+		}
+
+	case NET_IPV4_AUTOCONF_ANNOUNCE:
+		if (iface->ipv4.announce_cnt <=
+		    IPV4_AUTOCONF_ANNOUNCE_NUM - 1) {
+			net_ipaddr_copy(&iface->ipv4.current_ip,
+					&iface->ipv4.requested_ip);
+			send_announcement(iface);
+			break;
+		}
+
+		/* success, add new IPv4 address */
+		if (!net_if_ipv4_addr_add(iface,
+					  &iface->ipv4.requested_ip,
+					  NET_ADDR_AUTOCONF, 0)) {
+			NET_DBG("Failed to add IPv4 addr to iface %p", iface);
+			return;
+		}
+
+		iface->ipv4.state = NET_IPV4_AUTOCONF_CONFIGURED;
+		break;
+	default:
+		break;
+	}
+}
+
+static inline uint32_t ipv4_autoconf_get_timeout(struct net_if *iface)
+{
+	switch (iface->ipv4.state) {
+	case NET_IPV4_AUTOCONF_PROBE:
+		if (iface->ipv4.conflict_cnt >=
+		    IPV4_AUTOCONF_MAX_CONFLICTS) {
+			NET_DBG("rate limiting\n");
+			return IPV4_AUTOCONF_RATE_LIMIT_INTERVAL *
+			       MSEC_PER_SEC;
+		} else if (iface->ipv4.probe_cnt ==
+			   IPV4_AUTOCONF_PROBE_NUM) {
+			return IPV4_AUTOCONF_ANNOUNCE_INTERVAL *
+			       MSEC_PER_SEC;
+		}
+		NET_DBG("delay 1s\n");
+		return IPV4_AUTOCONF_PROBE_WAIT *
+		       CONFIG_SYS_CLOCK_TICKS_PER_SEC +
+		       (sys_rand32_get() % MSEC_PER_SEC);
+	case NET_IPV4_AUTOCONF_ANNOUNCE:
+		return IPV4_AUTOCONF_ANNOUNCE_INTERVAL *
+		       MSEC_PER_SEC;
+	default:
+		break;
+	}
+
+	return 0;
+}
+
+static void send_probe(struct net_if *iface)
+{
+	struct net_pkt *probe = prepare_arp(iface);
+
+	if (!probe) {
+		NET_DBG("Failed to prepare probe %p", iface);
+		return;
+	}
+
+	NET_DBG("Probing");
+	net_if_queue_tx(iface, probe);
+	iface->ipv4.probe_cnt++;
+	iface->ipv4.state = NET_IPV4_AUTOCONF_PROBE;
+
+	k_delayed_work_init(&iface->ipv4.ipv4_autoconf_timeout,
+			    ipv4_autoconf_timeout);
+	k_delayed_work_submit(&iface->ipv4.ipv4_autoconf_timeout,
+			      ipv4_autoconf_get_timeout(iface));
+}
+
+static void send_announcement(struct net_if *iface)
+{
+	struct net_pkt *announce = prepare_arp(iface);
+
+	if (!announce) {
+		NET_DBG("Failed to prepare announcement %p", iface);
+		return;
+	}
+
+	NET_DBG("Announcing");
+	net_if_queue_tx(iface, announce);
```

Comment:
I wonder if we could let the ethernet.c to actually send the packet here. So could we call net_send_data() here instead, which would properly do various checks etc before actually sending the pkt?

Reply:
I fully agree here, and I don't like to see L2 specific code being cold in L3 (net_arp_prepare).
The other way round would be fine: ethernet could check if ipv4 autoconf is on going, and thus react accordingly with arp. (setting up the proper ipv4 address in arp request etc..)